overlord/hookstate/context.go

```diff
@@ -127,3 +159,52 @@ func (c *Context) Get(key string, value interface{}) error {
 
 	return nil
 }
+
+// State returns the state contained within the context
+func (c *Context) State() *state.State {
+	return c.task.State()
+}
+
+// Cached returns the cached value associated with the provided key. It returns
+// nil if there is no entry for key. Note that the context needs to be locked
+// and unlocked by the caller.
+func (c *Context) Cached(key interface{}) interface{} {
+	c.reading()
+
+	return c.cache[key]
+}
+
+// Cache associates value with key. The cached value is not persisted. Note that
+// the context needs to be locked/unlocked by the caller.
+func (c *Context) Cache(key, value interface{}) {
+	c.writing()
+
+	c.cache[key] = value
+}
+
+// OnDone requests the provided function to be run once the context knows it's
+// complete. This can be called multiple times; each function will be called in
+// the order in which they were added. Note that the context needs to be locked
+// and unlocked by the caller.
+func (c *Context) OnDone(f func() error) {
+	c.writing()
+
+	c.onDone = append(c.onDone, f)
+}
+
+// Done is called to notify the context that its hook has exited successfully.
+// It will call all of the functions added in OnDone (even if one of them
+// returns an error) and will return the first error encountered. Note that the
+// context needs to be locked/unlocked by the caller.
+func (c *Context) Done() error {
```

Comment:
Yeah I went back and forth on that, but in the end since every other data-related function here required the caller to hold the lock, I figured it would be less error-prone to maintain consistency. Super easy change if we end up not liking that, though.